# Submission for HDSC-21 Stage A assesment

## Dataset description

'Record_id' : record id

'Utility_id_ferc1': Utility id assigned by the FERC

'Report_year': year of report

'Plant_name_ferc1': the name of the plant

'Fuel_type_code_pudl': the type of fuel

'Fuel_unit': the unit of fuel

'Fuel_qty_burned': the quantity of fuel burned

'Fuel_mmbtu_per_unit': the measure of energy per unit

'fuel_cost_per_unit_burned': the fuel cost per unit burned

'Fuel_cost_per_unit_delivered': the cost of fuel delivered per unit

'fuel_cost_per_mmbtu': the cost of fuel per mmbtu

In [26]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [8]:
fuel_df = pd.read_csv('../Dataset/fuel_ferc1.csv')
fuel_df.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [87]:
fuel_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     29523 non-null  object 
 1   utility_id_ferc1              29523 non-null  int64  
 2   report_year                   29523 non-null  int64  
 3   plant_name_ferc1              29523 non-null  object 
 4   fuel_type_code_pudl           29523 non-null  object 
 5   fuel_unit                     29343 non-null  object 
 6   fuel_qty_burned               29523 non-null  float64
 7   fuel_mmbtu_per_unit           29523 non-null  float64
 8   fuel_cost_per_unit_burned     29523 non-null  float64
 9   fuel_cost_per_unit_delivered  29523 non-null  float64
 10  fuel_cost_per_mmbtu           29523 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


## Question 1

In [77]:
A = [1,2,3,4,5,6]
B = [13, 21, 34]
A.extend(B)
A

[1, 2, 3, 4, 5, 6, 13, 21, 34]

## Question 2

In [78]:
np.array([[1,0,0],[0,1,0],[0,0,1]])
#0r
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

## Question 3

In [54]:
#Question 3 solution
#groupby fuel code and find the average
fuel_df.groupby('fuel_type_code_pudl').mean()

utility_id_ferc1  report_year  fuel_qty_burned  \
fuel_type_code_pudl                                                   
coal                       115.637534  2004.623143     1.903473e+06   
gas                        124.122062  2005.978060     4.843183e+06   
nuclear                    107.397311  2002.970660     5.454838e+06   
oil                        113.236235  2007.195933     6.311677e+04   
other                      110.467066  2001.700599     2.948137e+06   
waste                      154.253968  2005.650794     1.171914e+05   

                     fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
fuel_type_code_pudl                                                   
coal                           20.762780                  67.421830   
gas                             1.021913                  13.659397   
nuclear                        17.582120                4955.157002   
oil                             5.814377                 168.877086   
other                           0.572752                  18.253856   
waste                           0.341163                  19.518122   

                     fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
fuel_type_code_pudl                                                     
coal                                   116.951141             1.787190  
gas                                     12.095172             5.408876  
nuclear                              28616.915039             1.716559  
oil                                    313.907691            12.698732  
other                                   16.871485           155.084910  
waste                                   18.413052           822.709937

## Question 4

In [53]:
#Question 4 solution
#use the describe method to generate summary statistics
fuel_df['fuel_mmbtu_per_unit'].describe()

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

## Question 5

In [56]:
#Question 5 solution
skew = fuel_df['fuel_qty_burned'].skew()
kurt = fuel_df['fuel_qty_burned'].kurtosis()
print(skew, kurt)

15.851495469109503 651.3694501337732


## Question 6

In [58]:
fuel_df.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [61]:
fuel_df.describe()

utility_id_ferc1   report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count      29523.000000  29523.000000     2.952300e+04         29523.000000   
mean         118.601836   2005.806050     2.622119e+06             8.492111   
std           74.178353      7.025483     9.118004e+06            10.600220   
min            1.000000   1994.000000     1.000000e+00             0.000001   
25%           55.000000   2000.000000     1.381700e+04             1.024000   
50%          122.000000   2006.000000     2.533220e+05             5.762694   
75%          176.000000   2012.000000     1.424034e+06            17.006000   
max          514.000000   2018.000000     5.558942e+08           341.260000   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count               29523.000000                  2.952300e+04   
mean                  208.649031                  9.175704e+02   
std                  2854.490090                  6.877593e+04   
min                  -276.080000                 -8.749370e+02   
25%                     5.207000                  3.778500e+00   
50%                    26.000000                  1.737100e+01   
75%                    47.113000                  4.213700e+01   
max                139358.000000                  7.964521e+06   

       fuel_cost_per_mmbtu  
count         29523.000000  
mean             19.304354  
std            2091.540939  
min             -41.501000  
25%               1.940000  
50%               4.127000  
75%               7.745000  
max          359278.000000

The dataset has a total of 29523 rows and the missing rows of fuel_unit are 180

In [67]:
perct = (180/29523) * 100
print(perct)
round(perct, 3)

0.609694136774718


0.61

## Question 8


In [69]:
fuel_df.corr()['fuel_cost_per_unit_burned'].sort_values()

utility_id_ferc1               -0.037863
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

The 2nd and 3rd lowest correlations with fuel cost per unit burned are: fuel_mmbtu_per_unit (0.010034) and fuel_cost_per_unit_delivered(0.11007).

Negative sign indicates negative correlation type. Correlation values are copared in magnitude.

## Question 9

In [85]:
fuel_df['fuel_type_code_pudl'] = fuel_df['fuel_type_code_pudl'].astype(str)

Divide the data set in year 1998 and 1994 where fuel tpe is only coal. Then take the average of the fuel burned per unit cost for each year and calculate the percentage differance.

In [98]:
#dividing into 1994 and 1998
y_1994 = fuel_df.copy()[(fuel_df['report_year'] == 1994) & (fuel_df['fuel_type_code_pudl'] == 'coal')]
y_1998 = fuel_df.copy()[(fuel_df['report_year'] == 1998) & (fuel_df['fuel_type_code_pudl'] == 'coal')]

In [102]:
#calculation for percentage change
difference = y_1998['fuel_cost_per_unit_burned'].sum() - y_1994['fuel_cost_per_unit_burned'].sum()
change = (difference / y_1998['fuel_cost_per_unit_burned'].sum()) * 100
change

-25.893298748163957

## Question 10

In [75]:
#Question 10 solution
#sort values in descending order to get the max value
fuel_df.groupby('report_year')['fuel_cost_per_unit_delivered'].mean().sort_values(ascending=False)

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: fuel_cost_per_unit_delivered, dtype: float64